README.txt
plone.recipe.zope2install
=========================

Options
-------

To specify which Zope 2 to use, use one of the following options:

`url`
    The URL to a tarball to use for the Zope 2 installation.

`svn`
    The URL for an subversion checkout to use for the Zope 2 installation.

`location`
    The path to a local, existing Zope 2 installation. Note: For this to work
    Zope must have been built with the same Python used to run buildout.

`fake-zope-eggs`
    If set to true, add fake egg links to Zope 3 libraries, so that setuptools
    can see and use them for dependencies lookup. Enabled by default since
    version 3.0 of this recipe.

`additional-fake-eggs`
    Specify an optional list of additional fake eggs. Only include packages
    which are available on the Python path.

    You can also specify an explicit version to fake for an egg. For example::

        additional-fake-eggs =
            ZODB3 = 3.7.1
            zope.annotation = 3.3.2

    Otherwise the faked eggs will always have version 0.0.
    
    The default value includes Acquisition, ClientForm, DateTime, docutils,
    ExtensionClass, mechanize, Persistence, pytz, RestrictedPython,
    tempstorage, ZConfig, zLOG, zodbcode, ZODB3, zdaemon and Zope2.

`skip-fake-eggs`
    Specify an optional list of packages, for whom no fake egg is created.
    This allows to pull in new versions of some of the Zope packages via
    normal version requirements.

`smart-recompile`
    Will not recompile Zope if it finds .so or .pyd files. This means
    you can move your buildout around and speed up builds.

`python`
    Specify a section that configures another executable than the one used by 
    ``bin/buildout``.  For example::

        [zope2]
        recipe = plone.recipe.zope2install
        ...
        python = python2.4
        
        [python2.4]
        executable = ${buildout:directory}/parts/python/bin/python

If you use many buildouts with the same Zope 2 version, then you can add
"zope-directory" in the "buildout" section in your ~/.buildout/default.cfg
file like this::

    [buildout]
    zope-directory = /home/me/buildout/zope

For installations from a tarball that directory will be used instead of the
parts directory in your buildout. Each version of Zope will get it's own
directory but if it's already installed the existing one will be reused.

Exported variables
------------------

The following variables are set by this recipe and can be used in other parts.

`location`
    The path to the Zope installation root.

Reporting bugs or asking questions
----------------------------------

We have a shared bug tracker and help desk on Launchpad:
https://bugs.launchpad.net/collective.buildout/


CHANGES.txt
Change History
==============

3.4 (unreleased)
----------------

- Nothing changed yet.


3.3 (2010-09-01)
----------------

- Add ``python`` option to be able to specify other python than
  ``buildout:executable``
  [gotcha]


3.2 (2009-08-07)
----------------

- Refactor egg faking; fix skipped eggs to avoid to remove skipped
  develop eggs
  [gotcha]


3.1 (2009-06-23)
----------------

- Create download-cache directory when not provided by buildout
  [gotcha, jfroche]

- Change interpreter name in buildout to avoid name clash with an existing interpreter
  [gotcha, jfroche]

3.0 (2009-02-26)
----------------

- Changed the default for fake eggs creation to true and added a number of
  additional fake eggs to the default.
  [hannosch]

2.6 (2008-12-12)
----------------

- Change the fake eggs creation, to fix bug related to egg dependencies
  in develop-eggs.
  [jfroche]

- Added the smart-recompile option.
  [tarek]

- Ignore exit status of compilezpy.py since it returns an
  exit status > 0 for even a single compile failure.
  [smcmahon]

2.5 (2008-11-15)
----------------

- Add compilation of .py files into .pyc using Zope's compilezpy.py.
  [smcmahon]


2.4 (2008-11-04)
----------------

- Allow to use an existing Zope installation by specifying its `location`.
  [stefan]

- Prevent update from ever messing with a shared Zope installation.
  [fschulze]


2.3 (2008-08-27)
----------------

- stderr in svn update is no longer filtered but printed on console.
  [icemac]


2.2 (2008-06-07)
----------------

- Updated `fake-zope-eggs` handling, to be called when updating, so changes
  to `skip-fake-eggs` and add `additional-fake-eggs` are respected.
  [hannosch]

- Added new `skip-fake-eggs` option, that allows to skip the creation of
  fake eggs for a list of packages, so you can for example require a newer
  `zope.tal` version.
  [hannosch]

- Include the Python version Zope is compiled with into the shared-zope name.
  This allows you to have the same Zope version compiled with for example
  Python 2.3 and 2.4 but still share it. The change means, you will have to
  recompile all Zope versions.
  [hannosch, fschulze]

- Added error message for 0-byte sized Zope tarball downloads.
  [hannosch, sidnei]


2.1 (2008-06-05)
----------------

- Fixed error in retracted 2.0 release.
  [hannosch]

- Explicitly set zip-safe flag.
  [hannosch]


2.0 (2008-06-05)
----------------

- Merged all the options from the affinitic.recipe.fakezope2eggs recipe into
  this recipe.
  [jfroche, hannosch]


1.2 (2008-01-05)
----------------

- Renamed download-directory to download-cache, which is the name that is used
  by zc.buildout itself.
  [hannosch]


1.1 (2007-12-21)
----------------

- Small documentation update. Added link to the bugtracker.
  [hannosch]


1.0 (2007-11-29)
----------------

- Added changes file and small doc layout fix.
  [fschulze, hannosch]


0.3 (2007-09-05)
----------------

- Fixed an error where the "shared-zope" option was accessed but not yet set.
  [fschulze]


0.2 (2007-09-05)
----------------

- Added option to share Zope installations.
  [fschulze]


0.1 (2007-02-26)
----------------

- Initial implementation.
  [hannosch]


src/plone/recipe/zope2install/__init__.py
##############################################################################
#
# Copyright (c) 2007 Zope Corporation and Contributors.
# All Rights Reserved.
#
# This software is subject to the provisions of the Zope Public License,
# Version 2.1 (ZPL).  A copy of the ZPL should accompany this distribution.
# THIS SOFTWARE IS PROVIDED "AS IS" AND ANY AND ALL EXPRESS OR IMPLIED
# WARRANTIES ARE DISCLAIMED, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
# WARRANTIES OF TITLE, MERCHANTABILITY, AGAINST INFRINGEMENT, AND FITNESS
# FOR A PARTICULAR PURPOSE.
#
##############################################################################
import os
import re
import shutil
import sys
import tempfile
import urllib2
import urlparse
import setuptools.archive_util

EGG_INFO_CONTENT = """Metadata-Version: 1.0
Name: %s
Version: %s
"""

DEFAULT_FAKE_EGGS = [
    'Acquisition',
    'ClientForm',
    'DateTime',
    'docutils',
    'ExtensionClass',
    'mechanize',
    'pytz',
    'RestrictedPython',
    'Persistence',
    'tempstorage',
    'ZConfig',
    'zLOG',
    'zodbcode',
    'ZODB3',
    'zdaemon',
    'Zope2',
]

# define which values are read as true
TRUEVALS = ('y', 'yes', 't', 'true', 'on', '1')
VERSION_0_0 = '0.0'


class FakeLibInfo(object):
    """
    a really simple to store informations about libraries to be faked
    as eggs.
    """
    version = ''
    name = ''

    def __init__(self, name, version=VERSION_0_0, skip=False):
        self.version = version
        self.name = name
        self.skip = skip

    def make(self, fakeEggsDirectory, developEggsDirectory):
        self.link = os.path.join(developEggsDirectory,
            '%s.egg-link' % self.name)
        self.libDirectory = os.path.join(fakeEggsDirectory, self.name)
        if self.skip:
            self.cleanup()
        else:
            self.makeEggInfoFile()
            self.makeDevelopEggLink()

    def cleanup(self):
        if os.path.isdir(self.libDirectory):
            shutil.rmtree(self.libDirectory)
        if os.path.exists(self.link) and self.linkIsFake():
            os.remove(self.link)

    def makeEggInfoFile(self):
        directory = self.libDirectory
        if not os.path.isdir(directory):
            os.mkdir(directory)
        eggInfoFile = os.path.join(directory, '%s.egg-info' % self.name)
        fd = open(eggInfoFile, 'w')
        fd.write(EGG_INFO_CONTENT % (self.name, self.version))
        fd.close()

    def makeDevelopEggLink(self):
        fd = open(self.link, 'w')
        fd.write("%s\n." % self.libDirectory)
        fd.close()

    def linkIsFake(self):
        fd = open(self.link)
        try:
            line = fd.readline()
            return line.startswith(self.libDirectory)
        finally:
            fd.close()


class Recipe:

    def __init__(self, buildout, name, options):
        self.buildout, self.name, self.options = buildout, name, options

        python = options.get('python', buildout['buildout']['python'])
        options['executable'] = buildout[python]['executable']
        self.location = options.get('location', None)
        self.svn = options.get('svn', None)
        self.url = options.get('url', None)
        assert self.location or self.svn or self.url

        if self.location is not None:
            # We have an existing Zope installation; use it.
            assert os.path.exists(self.location), \
                'No such file or directory: %s' % self.location
            options['location'] = self.location
            options['shared-zope'] = 'true'
        elif (self.svn is None and
            buildout['buildout'].get('zope-directory') is not None):
            # if we use a download, then we look for a directory with shared
            # Zope installations. TODO Sharing of SVN checkouts is not yet
            # supported
            _, _, urlpath, _, _, _ = urlparse.urlparse(self.url)
            fname = urlpath.split('/')[-1]
            # cleanup the name a bit
            for s in ('.tar', '.bz2', '.gz', '.tgz'):
                fname = fname.replace(s, '')
            # Include the Python version Zope is compiled with into the
            # download cache name, so you can have the same Zope version
            # compiled with for example Python 2.3 and 2.4 but still share it
            ver = sys.version_info[:2]
            pystring = 'py%s.%s' % (ver[0], ver[1])
            options['location'] = os.path.join(
                buildout['buildout']['zope-directory'],
                '%s-%s' % (fname, pystring))
            options['shared-zope'] = 'true'
        else:
            # put it into parts
            options['location'] = os.path.join(
                buildout['buildout']['parts-directory'],
                self.name)
        # We look for a download cache, where we put the downloaded tarball
        if buildout['buildout'].get('download-cache') is None:
            download_cache = os.path.join(buildout['buildout']['directory'],
                                          'downloads')
            if not os.path.isdir(download_cache):
                os.mkdir(download_cache)
            buildout['buildout'].setdefault('download-cache', download_cache)

        self.fake_zope_eggs = options.get('fake-zope-eggs', 'true')
        if self.fake_zope_eggs.lower() not in TRUEVALS:
            self.fake_zope_eggs = False
        else:
            self.fake_zope_eggs = True
        self.fake_eggs_folder = self.options.get('fake-eggs-folder',
                                                 'fake-eggs')

        skip_fake_eggs = self.options.get('skip-fake-eggs', '')
        self.skip_fake_eggs = [e for e in skip_fake_eggs.split('\n') if e]

    def _compiled(self, path):
        """returns True if the path is compiled"""
        for files, dirs, root in os.walk(path):
            for f in files:
                base, ext = os.path.splitext(f)
                if ext == '.c':
                    if sys.platform == 'win32':
                        compiled_ext = '.pyd'
                    else:
                        compiled_ext = '.so'
                    compiled = os.path.join(root,
                        '%s%s' % (base, compiled_ext))
                    if not os.path.exists(compiled):
                        return False
        return True

    def install(self):
        options = self.options
        print 'Zope2 is setup using : %s' % options['executable']
        location = options['location']
        download_dir = self.buildout['buildout']['download-cache']
        smart_recompile = options.get('smart-recompile') == 'true'

        if os.path.exists(location):
            # if the zope installation exists and is shared, then we are done
            # and don't return a path, so the shared installation doesn't get
            # deleted on uninstall
            if options.get('shared-zope') == 'true':
                # We update the fake eggs in case we have special skips or
                # additions
                if self.fake_zope_eggs:
                    print 'Creating fake eggs'
                    self.fakeEggs()
                return []
        else:
            smart_recompile = True

        if smart_recompile and os.path.exists(location):
            # checking if the c source where compiled.
            if self._compiled(location):
                if self.fake_zope_eggs:
                    print 'Creating fake eggs'
                    self.fakeEggs()
                return []

        # full installation
        if os.path.exists(location):
            shutil.rmtree(location)

        if self.svn:
            assert os.system('svn co %s %s' % (options['svn'], location)) == 0
        else:
            if not os.path.isdir(download_dir):
                os.mkdir(download_dir)

            _, _, urlpath, _, _, _ = urlparse.urlparse(self.url)
            tmp = tempfile.mkdtemp('buildout-' + self.name)
            try:
                fname = os.path.join(download_dir, urlpath.split('/')[-1])
                # Have we already downloaded the file
                if not os.path.exists(fname):
                    f = open(fname, 'wb')
                    try:
                        f.write(urllib2.urlopen(self.url).read())
                    except:
                        os.remove(fname)
                        raise
                    f.close()

                setuptools.archive_util.unpack_archive(fname, tmp)
                # The Zope tarballs have a Zope-<version> folder at the root
                # level, so we need to move that one into the right place.
                files = os.listdir(tmp)
                if len(files) == 0:
                    raise ValueError('Broken Zope tarball named %s' % fname)
                shutil.move(os.path.join(tmp, files[0]), location)
            finally:
                shutil.rmtree(tmp)

        os.chdir(location)
        assert os.spawnl(
            os.P_WAIT, options['executable'], options['executable'],
            'setup.py',
            'build_ext', '-i',
            ) == 0

        # compile .py files to .pyc;
        # ignore return status since compilezpy.py will return
        # an exist status of 1 for even a single failed compile.
        os.spawnl(
            os.P_WAIT, options['executable'], options['executable'],
            os.path.join(location, 'utilities', 'compilezpy.py'),
            'build_ext', '-i',
            )

        if self.fake_zope_eggs:
            print 'Creating fake eggs'
            self.fakeEggs()
        if self.url and options.get('shared-zope') == 'true':
            # don't return path if the installation is shared, so it doesn't
            # get deleted on uninstall
            return []
        return location

    def fakeEggs(self):
        self.makeFakeEggsDirectory()
        self.initAdditionalFakeEggs()
        self.harvestPotentialFakeEggs()

        developEggDir = self.buildout['buildout']['develop-eggs-directory']
        for libInfo in self.libsToFake:
            libInfo.make(self.fakeEggsDirectory, developEggDir)

    def makeFakeEggsDirectory(self):
        self.fakeEggsDirectory = os.path.join(
            self.buildout['buildout']['directory'], self.fake_eggs_folder)
        if not os.path.isdir(self.fakeEggsDirectory):
            os.mkdir(self.fakeEggsDirectory)

    def initAdditionalFakeEggs(self):
        self.libsToFake = []
        # We add all additional fake eggs
        additional = self.options.get('additional-fake-eggs', '')
        additional = [e for e in additional.split('\n') if e]
        additional_names = []
        # Build up a list of all fake egg names without a version spec
        for line in additional:
            if '=' in line:
                spec = line.strip().split('=')
                name = spec[0].strip()
            else:
                name = line.strip()
            additional_names.append(name)
        # Add defaults to the specified set if the egg is not specified
        # in the additional-fake-eggs option, so you can overwrite one of
        # the default eggs with one including a version spec
        for name in DEFAULT_FAKE_EGGS:
            if name not in additional_names:
                additional.append(name)
        for lib in additional:
            # 2 forms available:
            #  * additional-fake-eggs = myEgg
            #  * additional-fake-eggs = myEgg=0.4
            if '=' in lib:
                lib = lib.strip().split('=')
                eggName = lib[0].strip()
                version = lib[1].strip()
            else:
                eggName = lib.strip()
                version = VERSION_0_0
            skip = self.shouldSkipFakeEgg(eggName)
            libInfo = FakeLibInfo(eggName, version, skip)
            self.libsToFake.append(libInfo)

    def shouldSkipFakeEgg(self, name):
        return name in self.skip_fake_eggs

    def harvestPotentialFakeEggs(self):
        zope2Location = self.options['location']
        zopeLibZopeLocation = os.path.join(zope2Location, 'lib', 'python',
                                           'zope')
        zopeLibZopeAppLocation = os.path.join(zope2Location, 'lib', 'python',
                                              'zope', 'app')
        self.libsToFake += self._getInstalledLibs(zopeLibZopeLocation, 'zope')
        self.libsToFake += self._getInstalledLibs(zopeLibZopeAppLocation,
                                             'zope.app')

    def _getInstalledLibs(self, location, prefix):
        installedLibs = []
        for lib in os.listdir(location):
            name = '%s.%s' % (prefix, lib)
            if (os.path.isdir(os.path.join(location, lib)) and
                name not in [libInfo.name for libInfo in self.libsToFake]):
                # Only add the package if it's not yet in the list
                skip = self.shouldSkipFakeEgg(name)
                installedLibs.append(FakeLibInfo(name, skip=skip))
        return installedLibs

    def update(self):
        options = self.options
        location = options['location']
        if os.path.exists(location):
            # Don't do anything in offline mode
            if self.buildout['buildout'].get('offline') == 'true' or \
               self.buildout['buildout'].get('newest') == 'false':
                if self.fake_zope_eggs:
                    print 'Updating fake eggs'
                    self.fakeEggs()
                if options.get('shared-zope') == 'true':
                    return []
                return location

            # If we downloaded a tarball, we don't need to do anything while
            # updating, otherwise we have a svn checkout and should run
            # 'svn up' and see if there has been any changes so we recompile
            # the c extensions
            if self.location or self.url:
                if self.fake_zope_eggs:
                    print 'Updating fake eggs'
                    self.fakeEggs()
                if options.get('shared-zope') == 'true':
                    return []
                return location

            if (self._compiled(location) and
                options.get('smart-recompile') == 'true'):
                return location

            os.chdir(location)
            stdin, stdout, stderr = os.popen3('svn up')
            stdin.close()
            for line in stderr.readlines():
                sys.stderr.write(line)
            stderr.close()
            change = re.compile('[AUM] ').match
            for l in stdout:
                if change(l):
                    stdout.read()
                    stdout.close()
                    break
                else:
                    # No change, so all done
                    stdout.close()
                    return location

            assert os.spawnl(
                os.P_WAIT, options['executable'], options['executable'],
                'setup.py',
                'build_ext', '-i',
                ) == 0

            if self.fake_zope_eggs:
                print 'Updating fake eggs'
                self.fakeEggs()

        return location
